ComponentKit/HostingView/CKComponentHostingView.mm

```diff
@@ -40,108 +48,127 @@ - (instancetype)initWithFrame:(CGRect)frame
+  CKUnmountComponents(_mountedComponents);
 }
 
 #pragma mark - Layout
 
 - (void)layoutSubviews
 {
+  CKAssertMainThread();
   [super layoutSubviews];
   _containerView.frame = self.bounds;
 
-  if (_model && !CGRectIsEmpty(self.bounds)) {
-    [self _update];
-
-    if (![_lifecycleManager isAttachedToView]) {
-      [_lifecycleManager attachToView:_containerView];
+  if (!CGRectIsEmpty(self.bounds)) {
+    [self _updateComponentIfNeeded];
+    const CGSize size = self.bounds.size;
+    if (_layout.component != _component || !CGSizeEqualToSize(_layout.size, size)) {
+      _layout = [_component layoutThatFits:{size, size} parentSize:size];
     }
+    _mountedComponents = [CKMountComponentLayout(_layout, _containerView, _mountedComponents, nil) copy];
   }
 }
 
 - (CGSize)sizeThatFits:(CGSize)size
 {
-  CKSizeRange constrainedSize = [_sizeRangeProvider sizeRangeForBoundingSize:size];
-  CKComponentLayout layout = [_lifecycleManager layoutForModel:_model constrainedSize:constrainedSize context:_context];
-  return layout.size;
+  CKAssertMainThread();
+  [self _updateComponentIfNeeded];
+  const CKSizeRange constrainedSize = [_sizeRangeProvider sizeRangeForBoundingSize:size];
+  return [_component layoutThatFits:constrainedSize parentSize:constrainedSize.max].size;
 }
 
 #pragma mark - Accessors
 
 - (void)setModel:(id)model
 {
+  CKAssertMainThread();
   if (_model != model) {
     _model = model;
-    CKAssertNotNil(_model, @"Model can not be nil.");
-
-    [self setNeedsLayout];
+    [self _setComponentNeedsUpdate];
   }
 }
 
 - (void)setContext:(id<NSObject>)context
 {
+  CKAssertMainThread();
   if (_context != context) {
     _context = context;
-    [self setNeedsLayout];
+    [self _setComponentNeedsUpdate];
   }
 }
 
-- (UIView *)containerView
+- (const CKComponentLayout &)mountedLayout
 {
-  return _containerView;
+  return _layout;
 }
 
-#pragma mark - CKComponentLifecycleManagerDelegate
+#pragma mark - CKComponentStateListener
 
-- (void)componentLifecycleManager:(CKComponentLifecycleManager *)manager
-       sizeDidChangeWithAnimation:(const CKComponentBoundsAnimation &)animation
+- (void)componentScopeHandleWithIdentifier:(CKComponentScopeHandleIdentifier)globalIdentifier
+                            rootIdentifier:(CKComponentScopeRootIdentifier)rootIdentifier
+                     didReceiveStateUpdate:(id (^)(id))stateUpdate
+                     tryAsynchronousUpdate:(BOOL)tryAsynchronousUpdate
 {
-  [_delegate componentHostingViewDidInvalidateSize:self];
+  CKAssertMainThread();
+  _pendingStateUpdates.insert({globalIdentifier, stateUpdate});
+  [self _setComponentNeedsUpdate];
 }
 
 #pragma mark - Private
 
-- (void)_update
+- (void)_setComponentNeedsUpdate
 {
+  if (!_componentNeedsUpdate) { // Avoid thrashing delegate
+    _componentNeedsUpdate = YES;
+    [self setNeedsLayout];
+    [_delegate componentHostingViewDidInvalidateSize:self];
+  }
+}
+
+- (void)_updateComponentIfNeeded
```

Comment:
Yeah, `CKTransactionalComponentDataSource` does have such an API as well. I'll probably add it in a follow-on then.